3scale/APIcast
# Fix dev env on aarch64 hosts

Fixes multiple issues to build and run the development environment on aarch64-based host machines, caused by several incompatible dependencies.

- [x] build and run dev image targeting linux/amd64 platform
- [x] deps: [IPC::Run](https://metacpan.org/pod/IPC::Run) (preinstalled) - because otherwise Test-Nginx cannot be installed
- [x] deps: [lua-resty-jwt](https://luarocks.org/modules/cdbattags/lua-resty-jwt) v0.2.0 (installed with Rover) - because latest version (v0.2.3) [requires](https://github.com/cdbattags/lua-resty-jwt/blob/b3d5c085643fa95099e72a609c57095802106ff9/lua-resty-jwt-dev-0.rockspec#L19) incompatible lua-resty-openssl
- [x] deps: [jsonschema](https://luarocks.org/modules/membphis/jsonschema) v0.8 (installed with Rover) - because latest version (v0.9.8) [requires](https://github.com/api7/jsonschema/blob/0096f5e5bfdd210136a98e9d50799d86b9bd040c/rockspec/jsonschema-0.9.8-0.rockspec#L23) incompatible lrexlib-pcre.
- [x] deps: [luaossl](https://luarocks.org/modules/daurnimator/luaossl) 20200709 (preinstalled) - because latest (20220711) cannot be compiled with OpenSSL 1.1.1 on an aarch64 host ([ref](https://github.com/wahern/luaossl/issues/175#issuecomment-1310378633))
- [x] deps: deps: [nginx-lua-prometheus](https://luarocks.org/modules/knyar/nginx-lua-prometheus) v0.20181120 - to avoid _"module 'nginx.prometheus' not found"_ when running the tests
- [x] deps: [lua-resty-http](https://luarocks.org/modules/pintsized/lua-resty-http) v0.15 - to avoid `string` vs `table` type when deep invoking `is_fqdn` from `apicast.configuration_loader.remote_v1:download → resty.resolver.http:request_uri`
- [x] fix: test assertion of a falsy value passing no argument – no longer the same as passing `nil` since [luassert](https://luarocks.org/modules/lunarmodules/luassert) v1.9.0.

Closes #1375.

### Verification steps

(Optional) Build your own dev image:

```sh
make dev-build
```

Run the dev env:

```sh
make development IMAGE=apicast-development:latest
```

...or using a prebuilt image containing the fixes:

```sh
make development IMAGE=quay.io/guicassolato/3scale:apicast-dev-20221124
```

Within the container, install the dependencies:

```sh
make dependencies
```

After several minutes, installation should finish successfully:

```
Complete! Modules were bundled into /opt/app-root/src/vendor/cache
```

Run the tests:

```sh
make busted
# 1208 successes / 0 failures / 0 errors / 3 pending : 24.479539 seconds
```

```sh
make prove
# All tests successful.
# Files=85, Tests=3397, 1601 wallclock secs ( 1.52 usr  0.20 sys + 3998.34 cusr 406.44 csys = 4406.50 CPU)
# Result: PASS
# Success
```

### Future improvement

Ditch Lua Rover altogether and use LuaRocks directly instead to install the dependencies.
